### PR TITLE
[MRG+1] ENH: Optional 3d bar shading

### LINKED ISSUE #7683: Please add <shade = True> parameter to bar3d
Matplotlib 2.0.0rc2
from source

Currently matplotlib always shades bar3d faces. This can be confusing at times, but could be alleviated by including a "shade" parameter. Unfortunately I do not regularly use Github, so I don't know how to make Pull requests. I've included updated code withe the enhanced functionality, however:

    def bar3d(self, x, y, z, dx, dy, dz, color=None,
              zsort='average', shade = True, *args, **kwargs):
        '''
        Generate a 3D bar, or multiple bars.

        When generating multiple bars, x, y, z have to be arrays.
        dx, dy, dz can be arrays or scalars.

        *color* can be:

         - A single color value, to color all bars the same color.

         - An array of colors of length N bars, to color each bar
           independently.

         - An array of colors of length 6, to color the faces of the
           bars similarly.

         - An array of colors of length 6 * N bars, to color each face
           independently.

         When coloring the faces of the boxes specifically, this is
         the order of the coloring:

          1. -Z (bottom of box)
          2. +Z (top of box)
          3. -Y
          4. +Y
          5. -X
          6. +X

        Keyword arguments are passed onto
        :func:`~mpl_toolkits.mplot3d.art3d.Poly3DCollection`
        '''
        had_data = self.has_data()

        if not cbook.iterable(x):
            x = [x]
        if not cbook.iterable(y):
            y = [y]
        if not cbook.iterable(z):
            z = [z]

        if not cbook.iterable(dx):
            dx = [dx]
        if not cbook.iterable(dy):
            dy = [dy]
        if not cbook.iterable(dz):
            dz = [dz]

        if len(dx) == 1:
            dx = dx * len(x)
        if len(dy) == 1:
            dy = dy * len(y)
        if len(dz) == 1:
            dz = dz * len(z)

        if len(x) != len(y) or len(x) != len(z):
            warnings.warn('x, y, and z must be the same length.')

        # FIXME: This is archaic and could be done much better.
        minx, miny, minz = 1e20, 1e20, 1e20
        maxx, maxy, maxz = -1e20, -1e20, -1e20

        polys = []
        for xi, yi, zi, dxi, dyi, dzi in zip(x, y, z, dx, dy, dz):
            minx = min(xi, minx)
            maxx = max(xi + dxi, maxx)
            miny = min(yi, miny)
            maxy = max(yi + dyi, maxy)
            minz = min(zi, minz)
            maxz = max(zi + dzi, maxz)

            polys.extend([
                ((xi, yi, zi), (xi + dxi, yi, zi),
                    (xi + dxi, yi + dyi, zi), (xi, yi + dyi, zi)),
                ((xi, yi, zi + dzi), (xi + dxi, yi, zi + dzi),
                    (xi + dxi, yi + dyi, zi + dzi), (xi, yi + dyi, zi + dzi)),

                ((xi, yi, zi), (xi + dxi, yi, zi),
                    (xi + dxi, yi, zi + dzi), (xi, yi, zi + dzi)),
                ((xi, yi + dyi, zi), (xi + dxi, yi + dyi, zi),
                    (xi + dxi, yi + dyi, zi + dzi), (xi, yi + dyi, zi + dzi)),

                ((xi, yi, zi), (xi, yi + dyi, zi),
                    (xi, yi + dyi, zi + dzi), (xi, yi, zi + dzi)),
                ((xi + dxi, yi, zi), (xi + dxi, yi + dyi, zi),
                    (xi + dxi, yi + dyi, zi + dzi), (xi + dxi, yi, zi + dzi)),
            ])

        facecolors = []
        if color is None:
            color = [self._get_patches_for_fill.get_next_color()]

        if len(color) == len(x):
            # bar colors specified, need to expand to number of faces
            for c in color:
                facecolors.extend([c] * 6)
        else:
            # a single color specified, or face colors specified explicitly
            facecolors = list(mcolors.to_rgba_array(color))
            if len(facecolors) < len(x):
                facecolors *= (6 * len(x))

        normals = self._generate_normals(polys)
	if shade:
		sfacecolors = self._shade_colors(facecolors, normals)
	else:
		sfacecolors = facecolors
        col = art3d.Poly3DCollection(polys,
                                     zsort=zsort,
                                     facecolor=sfacecolors,
                                     *args, **kwargs)
        self.add_collection(col)

        self.auto_scale_xyz((minx, maxx), (miny, maxy), (minz, maxz), had_data)

        return col

=====================
[bar3d_snippet.txt](https://github.com/matplotlib/matplotlib/files/672211/bar3d_snippet.txt)

### PR DESCRIPTION
**Pull request by phobson**
Closes #7683

### CODE REVIEW

File: examples/mplot3d/plot_3d_bars.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+# setup the figure and axes
+fig = plt.figure(figsize=(8, 3))
+ax1 = fig.add_subplot(121, projection='3d')
+ax2 = fig.add_subplot(122, projection='3d')
+
+# fake data
+_x = np.arange(4)
+_y = np.arange(5)
+_xx, _xx = np.meshgrid(_x, _y)
```
Comment (WeatherGod) [inline examples/mplot3d/plot_3d_bars.py]: typo here, should be `_xx, _yy = ...`

File: examples/mplot3d/plot_3d_bars.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+x, y = _xx.ravel(), _yy.ravel()
+
+top = x + y
+bottom = np.zeros_like(top)
+width = depth = 1
+
+ax1.bar3d(x, y, bottom, width, depth, top, shade=True)
```
Comment (WeatherGod) [inline examples/mplot3d/plot_3d_bars.py]: simplify to `ax1.bar()` instead of `ax1.bar3d()`

File: examples/mplot3d/plot_3d_bars.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ax1.set_title('Shaded')
+
+ax2.bar3d(x, y, bottom, width, depth, top, shade=False)
+ax2.set_title('Not Shaded')
+
+fig.canvas.draw()
```
Comment (WeatherGod) [inline examples/mplot3d/plot_3d_bars.py]: should be `plt.show()`

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 4]
```diff
@@ -2418,39 +2420,64 @@ def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
...
-         - An array of colors of length N bars, to color each bar
-           independently.
-
-         - An array of colors of length 6, to color the faces of the
-           bars similarly.
+              zsort='average', shade=True, *args, **kwargs):
+        """Generate a 3D barplot.
+
+        This method creates three dimensional barplot where the width,
+        depth, height, and color of the bars can all ne uniquely set.
```
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: typo: "ne" --> "be"

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 5]
```diff
@@ -2418,39 +2420,64 @@ def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
...
+              2. +Z (top of box)
+              3. -Y
+              4. +Y
+              5. -X
+              6. +X
+
+        zsort : str, optional
+            The z-axis sorting scheme passed onto
+            :func:`~mpl_toolkits.mplot3d.art3d.Poly3DCollection`
+        shape : bool, optional
```
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: typo: 'shape' --> 'shade'

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 6]
```diff
@@ -2522,7 +2549,11 @@ def bar3d(self, x, y, z, dx, dy, dz, color=None,
                 facecolors *= (6 * len(x))
 
         normals = self._generate_normals(polys)
```
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: This can also go into the if clause, since the normals aren't needed if there aren't any shading

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 7]
```diff
@@ -20,16 +20,37 @@ def test_bar3d():
...
+def test_bar3d_dflt_actually3d():
+    fig = plt.figure()
+    ax = fig.add_subplot(111, projection='3d')
+    x = np.arange(4)
+    y = np.arange(5)
+    x2d, y2d = np.meshgrid(x, y)
+    x2d, y2d = x2d.ravel(), y2d.ravel()
+    z = x2d + y2d
+    ax.bar3d(x2d, y2d, x2d * 0, 1, 1, z, shade=True)
+    fig.canvas.draw()
```
Comment (WeatherGod) [inline lib/mpl_toolkits/tests/test_mplot3d.py]: remove the unneeded `draw()`

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 8]
```diff
@@ -20,16 +20,37 @@ def test_bar3d():
...
 
 
-@cleanup
-def test_bar3d_dflt_smoke():
+@image_comparison(
+    baseline_images=['bar3d_shaded'],
+    remove_text=True,
+    extensions=['png']
+)
+def test_bar3d_dflt_actually3d():
```
Comment (WeatherGod) [inline lib/mpl_toolkits/tests/test_mplot3d.py]: I prefer test names and baseline image names that are the same or very nearly the same. It makes it easy to match a bad image file with the test in the code.

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 9]
```diff
@@ -20,16 +20,37 @@ def test_bar3d():
...
+    ax.bar3d(x2d, y2d, x2d * 0, 1, 1, z, shade=True)
+    fig.canvas.draw()
+
+
+@image_comparison(
+    baseline_images=['bar3d_notshaded'],
+    remove_text=True,
+    extensions=['png']
+)
+def test_bar3d_dflt_actually3d_no_shade ():
```
Comment (WeatherGod) [inline lib/mpl_toolkits/tests/test_mplot3d.py]: extraneous space between function name and parens.

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 10]
```diff
@@ -20,16 +20,37 @@ def test_bar3d():
...
     fig = plt.figure()
     ax = fig.add_subplot(111, projection='3d')
     x = np.arange(4)
     y = np.arange(5)
     x2d, y2d = np.meshgrid(x, y)
     x2d, y2d = x2d.ravel(), y2d.ravel()
     z = x2d + y2d
-    ax.bar3d(x2d, y2d, x2d * 0, 1, 1, z)
+    ax.bar3d(x2d, y2d, x2d * 0, 1, 1, z, shade=False)
     fig.canvas.draw()
```
Comment (WeatherGod) [inline lib/mpl_toolkits/tests/test_mplot3d.py]: go ahead and take it out here, too

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 11]
```diff
@@ -20,16 +20,37 @@ def test_bar3d():
         ax.bar(xs, ys, zs=z, zdir='y', color=cs, alpha=0.8)
 
 
-@cleanup
-def test_bar3d_dflt_smoke():
```
Comment (WeatherGod) [inline lib/mpl_toolkits/tests/test_mplot3d.py]: heh... I wonder why this was a smoke test?

File: examples/mplot3d/plot_3d_bars.py (hunk 1 above)
Comment (phobson) [inline examples/mplot3d/plot_3d_bars.py]: 👍 done

File: examples/mplot3d/plot_3d_bars.py (hunk 2 above)
Comment (phobson) [inline examples/mplot3d/plot_3d_bars.py]: I don't understand this one. I'm trying to draw 3-dimensional bars. I thought `bar` would draw 2-D bars on the 3D Axes object.

File: examples/mplot3d/plot_3d_bars.py (hunk 3 above)
Comment (phobson) [inline examples/mplot3d/plot_3d_bars.py]: 👍

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 6 above)
Comment (phobson) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Ahh. Thanks for noticing that.

File: lib/mpl_toolkits/tests/test_mplot3d.py (hunk 8 above)
Comment (phobson) [inline lib/mpl_toolkits/tests/test_mplot3d.py]: fair enough. I'll change them.

**Comment by WeatherGod** [comment]
Nope. Look at axes3d.py. You can't do 2d bar() with an Axes3d object. It
overrides the 2d version. The 3d part of many of the methods and attributes
are anachronisms and are only kept for back-compat.

On Jan 1, 2017 5:08 PM, "Paul Hobson" <notifications@github.com> wrote:

*@phobson* commented on this pull request.
------------------------------

In examples/mplot3d/plot_3d_bars.py
<https://github.com/matplotlib/matplotlib/pull/7691>:

> +# setup the figure and axes
+fig = plt.figure(figsize=(8, 3))
+ax1 = fig.add_subplot(121, projection='3d')
+ax2 = fig.add_subplot(122, projection='3d')
+
+# fake data
+_x = np.arange(4)
+_y = np.arange(5)
+_xx, _xx = np.meshgrid(_x, _y)
+x, y = _xx.ravel(), _yy.ravel()
+
+top = x + y
+bottom = np.zeros_like(top)
+width = depth = 1
+
+ax1.bar3d(x, y, bottom, width, depth, top, shade=True)

I don't understand this one. I'm trying to draw 3-dimensional bars. I
thought bar would draw 2-D bars on the 3D Axes object.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/matplotlib/matplotlib/pull/7691>, or mute the thread

**Comment by phobson** [comment]
Wait, I take it back, I'll still confused.

In axes3d.py, I see:

```python
    def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
        '''
        Add 2D bar(s).
        ...
        '''
```

And then:

```python
    def bar3d(self, x, y, z, dx, dy, dz, color=None,
              zsort='average', shade=True, *args, **kwargs):
        """Generate a 3D barplot.
        ...
        """
```

These seem like two very different and current methods. In the example that I created I wanted to demonstrate the difference that using shading makes, so 3D bars need to be generated in both cases.

The plot looks like this:

plot_3d_bars copy

Perhaps this would be clarified if the example was renamed to `plot_bars3d`

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 12]
```diff
@@ -2418,39 +2420,64 @@ def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
...
+        This method creates three dimensional barplot where the width,
+        depth, height, and color of the bars can all be uniquely set.
+
+        Parameters
+        ----------
+        x, y, z : array-like
+            The coordinates of the anchor point of the bars.
+
+        dx, dy, dz : scalar or array-like
+            The width, depth, and height of the bars, resprectively.
```
Comment (NelleV) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: typo: respectively.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 13]
```diff
@@ -2418,39 +2420,64 @@ def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
...
+              2. +Z (top of box)
+              3. -Y
+              4. +Y
+              5. -X
+              6. +X
+
+        zsort : str, optional
+            The z-axis sorting scheme passed onto
+            :func:`~mpl_toolkits.mplot3d.art3d.Poly3DCollection`
+        shade : bool, optional
```
Comment (NelleV) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: what's the default?

**Comment by NelleV** [comment]
Thanks for the example and the docstring update. It looks great!

**Comment by NelleV** [comment]
I hadn't seen the previous comments. I can have a look at it if you'd like

**Comment by QuLogic** [comment]
Any chance you will get back to this soon @phobson?

**Comment by phobson** [comment]
@QuLogic happy to rebase, but I'm still unclear what @WeatherGod meant with his comments.

**Comment by phobson** [comment]
@QuLogic rebased and squashed everything.
@WeatherGod could you take a second look at this. I'm still not sure what meant [by this](https://github.com/matplotlib/matplotlib/pull/7691#discussion_r94182463)

Despite the name of the example, `ax.bar` puts (2D bars in a 3D space)[http://matplotlib.org/devdocs/examples/mplot3d/bars3d_demo.html] while `ax.bar3d` draws [actual prisms](http://matplotlib.org/devdocs/examples/mplot3d/hist3d_demo.html)

File: examples/mplot3d/plot_3d_bars.py (hunk 2 above)
Comment (WeatherGod) [inline examples/mplot3d/plot_3d_bars.py]: ah, sorry, my reply apparently never got sent. For Axes3D objects, `ax.bar()` is the alias of `ax.bar3d()`. The `bar3d()` is an anachronism, much like `set_xlim3d()`.

Comment (WeatherGod) [inline examples/mplot3d/plot_3d_bars.py]: I take it all back. I could have sworn we separated the flat bar stuff out. You are right.

**Comment by WeatherGod** [comment]
We still have the other things I have noticed to fix, though.

**Comment by phobson** [comment]
@WeatherGod understood. I'm looking for clarity on what they might be.

**Comment by phobson** [comment]
@WeatherGod oops, sorry. I didn't see your other inline comments 🐑 

I took care of them and force-pushed

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 14]
```diff
@@ -2409,39 +2409,64 @@ def bar(self, left, height, zs=0, zdir='z', *args, **kwargs):
...
+              4. +Y
+              5. -X
+              6. +X
+
+        zsort : str, optional
+            The z-axis sorting scheme passed onto
+            :func:`~mpl_toolkits.mplot3d.art3d.Poly3DCollection`
+        shape : bool, optional (default = True)
+            When true, the dark sides of the bars (relative to the
+            plot's source of light), shaded.
```
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: typo: shape --> shade

Also, the description is missing a verb. Additionally, I think there is supposed to be a blank line between the zsort entry and the shade entry?

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 15]
```diff
@@ -2511,9 +2536,13 @@ def bar3d(self, x, y, z, dx, dy, dz, color=None,
             facecolors = list(mcolors.to_rgba_array(color))
             if len(facecolors) < len(x):
                 facecolors *= (6 * len(x))
+                normals = self._generate_normals(polys)
```
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: why is this here? wouldn't it make a lot more sense if it was in the `if shade` block?

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 14 above)
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Still missing a verb for this sentence, and possibly a blank line between the zsort entry and the shade entry.

**Comment by WeatherGod** [comment]
Ah, I almost completely forgot... we need a "what's new" entry for this.

**Comment by NelleV** [comment]
@phobson Do you think you'll have time to write this what's new entry?

File: doc/users/whats_new/toggle_3d_bar_shading.rst [hunk 16]
```diff
@@ -0,0 +1,28 @@
+Users can now toggle shading in 3D bar plots
+--------------------------------------------
+
+A new ``shade`` parameter has been added the 3D bar plotting method.
+The default behavior remains to the shade the bars, but now users
```
Comment (NelleV) [inline doc/users/whats_new/toggle_3d_bar_shading.rst]: I think it should be "to shade the bars"

Comment (phobson) [inline doc/users/whats_new/toggle_3d_bar_shading.rst]: gahhh i read that like 5 times. i'll fix :)

Comment (phobson) [inline doc/users/whats_new/toggle_3d_bar_shading.rst]: (shakes fist at self)

File: doc/users/whats_new/toggle_3d_bar_shading.rst [hunk 17]
```diff
@@ -0,0 +1,28 @@
+Users can now toggle shading in 3D bar plots
+--------------------------------------------
+
+A new ``shade`` parameter has been added the 3D bar plotting method.
+The default behavior remains to shade the bars, but now users
+have to the option of setting ``shade`` to ``False``.
```
Comment (WeatherGod) [inline doc/users/whats_new/toggle_3d_bar_shading.rst]: extraneous "to"

**Comment by phobson** [comment]
ping @WeatherGod
